MeshTangents : Added different modes for tangent calculation.

stefanfeess opened the pull request:
This adds the new Cortex MeshAlgoTangents computation methods from https://github.com/ImageEngine/cortex/pull/925.
We can now compute tangents not only based on UVs but also from the first edge of the vertex, the average of the fist two edges or the primitive centroid, make them orthogonal and flip the handedness.

### Dependencies ###

- Depends on: https://github.com/ImageEngine/cortex/pull/925

### Checklist ###

- [x] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/master/CONTRIBUTING.md).
- [x] I have updated the documentation, if applicable.
- [x] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [x] My code follows the Gaffer project's prevailing coding style and conventions.

mattigruener commented:
I've pointed Stefan to [the tests](https://github.com/GafferHQ/gaffer/blob/master/python/GafferSceneTest/MeshTangentsTest.py) that will probably need some fixing/amending.

andrewkaufman commented on python/GafferSceneTest/MeshTangentsTest.py:
```diff
@@ -91,6 +97,9 @@ def test( self ) :
 		meshTangents["filter"].setInput( pathFilter["out"] )
 
+		# Tangent Mode 0 (UV)
+		meshTangents['mode'].setValue(0)
```
should use the enum here and in the other tests

johnhaddon commented on python/GafferSceneTest/MeshTangentsTest.py:
```diff
@@ -52,10 +52,16 @@ def makeTriangleScene( self ) :
 		mesh = IECoreScene.MeshPrimitive( verticesPerFace, vertexIds, "linear", p )
 
+		mesh["N"] =  IECoreScene.PrimitiveVariable(
+			IECoreScene.PrimitiveVariable.Interpolation.FaceVarying,
```
Vertex probably? It seems the Cortex algo is assuming Vertex, although not yet checking for it...

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -36,51 +36,102 @@
 	"description",
 	"""
-	Adds surface tangent primitive variables to the mesh using the specified positions and UVSet. 
+	Adds surface tangent primitive variables to the mesh using the specified positions and UVSet.
```
This is out of date now we have a bunch of different modes. Let's update it.

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -36,51 +36,102 @@
+			"""
+			Name of the primitive variable which contains the normal vector jdata used calculate tangents & binormals.
+			""",
+			"layout:section", "Settings.Input",
```
If I understand correctly, we should add something to grey out the normal field when in UV mode?

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -36,51 +36,102 @@
 			"description",
 			"""
-			Name of the primitive variable which contains the position data used calculate tangents & binormals. 
+			Name of the primitive variable which contains the position data used calculate tangents & binormals.
```
"used *to* calculate"

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -36,51 +36,102 @@
+		"normal" : [
+			"description",
+			"""
+			Name of the primitive variable which contains the normal vector jdata used calculate tangents & binormals.
```
"normal vector jdata" -> "normals" 
"used calculate" -> "used to calculate"

johnhaddon commented on src/GafferScene/MeshTangents.cpp:
```diff
@@ -56,6 +56,9 @@ MeshTangents::MeshTangents( const std::string &name ) : SceneElementProcessor( n
 	addChild( new StringPlug( "uTangent", Plug::In, "uTangent" ) );
 	addChild( new StringPlug( "vTangent", Plug::In, "vTangent" ) );
 	addChild( new BoolPlug( "orthogonal", Plug::In, true ) );
+	addChild( new IntPlug( "mode", Plug::In, MeshAlgo::MeshTangentsMode::UV ) );
```
Specify min and max values corresponding to the min and max enum values.

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -36,51 +36,102 @@
+
+			"plugValueWidget:type", "GafferUI.PresetsPlugValueWidget",
+
+			"layout:index", 4,
```
It's a bit fragile specifying the order this way (despite being necessary sometimes), and it isn't respected when we auto-generate the docs. Let's just construct the plugs in the order we want, and order the accessors in the header correspondingly. That way we're presenting things in the same logical order wherever you look.

johnhaddon commented on python/GafferSceneUI/MeshTangentsUI.py:
```diff
@@ -36,51 +36,102 @@
-			Adjusts vTangent to be orthogonal to the uTangent. 
+			Name of the UV set primitive variable used to calculate uTangent & vTangent.
 			""",
+			"layout:section", "Settings.Input",
```
I like. Seems like this might be a good convention to adopt for all such nodes...

johnhaddon commented on src/GafferScene/MeshTangents.cpp:
```diff
@@ -151,12 +187,38 @@ IECore::ConstObjectPtr MeshTangents::computeProcessedObject( const ScenePath &pa
+	}
+	else
+	{
+		std::string e = boost::str( boost::format( "MeshTangents : Invalid MeshTangentMode \"%s\"." ) % mode );
```
If I recall correctly, you might be able to get away with it in boost::format, but %s isn't a logical specifier for an enum/int. If we enforce the appropriate min/max on the plug constructor, we won't need this block anyway...

stefanfeess commented:
Thanks for your review John! I pushed an update addressing your notes. Also I wondered if I should, for the UV mode, additionally output on `tangents` and `biTangents` so we can eventually deprecate the `uTangent` `vTangent` plugs. Or would that introduce to much overhead?

johnhaddon commented on include/GafferScene/MeshTangents.h:
```diff
@@ -79,6 +79,14 @@ class GAFFERSCENE_API MeshTangents : public SceneElementProcessor
 
 		IE_CORE_DECLARERUNTIMETYPEDEXTENSION( GafferScene::MeshTangents, MeshTangentsTypeId, SceneElementProcessor );
 
+    enum Mode
```
Tabs not spaces. Let's also put it higher up, so the reader has bumped into it before hitting the `modePlug()` - just one more clue what it's for.

stefanfeess replied:
Moved it up and converted to tabs. (...emacs hates me, and once in a while disables the c++ formatting layer) :D

stefanfeess commented:
Hey John,
I rebased onto your latest changes, fixed the whitespace, squashed everything down and, since I already touched that code, also added the post constructor to hide the uvTangent plugs.
